Mixed version compatibility detection

RanVaknin opened the pull request:
Following the release of a breaking change where a newer version of a service client was trying to hit a non-existent code path on an older version of core, we have identified a number of core classes that are considered "risky" to change.

When changed / added to, a "risky" core class can have a codepath that breaks customers using mixed SDK versions. For example, when newer service clients expect methods or functionality that older core versions don't provide, it causes runtime failures for customers who haven't upgraded all their SDK modules together.

This workflow detects changes to the following base classes that generated service code implements:
• `AwsRequest`
• `AwsResponse`
• `AwsResponseMetadata`
• `AwsServiceException`
• `SdkPojo`

The workflow uses git diff to detect any modifications to these specific base class files. If changes are found, it searches for new public method additions using pattern matching, and (attempts to) filter out obvious false positives like comments, string literals, and javadoc examples. When "risky" changes are detected, the workflow blocks the PR from merging until the team adds a `mixed-version-compatibility-reviewed label`.



Testing:
[tested on personal fork](https://github.com/RanVaknin/aws-sdk-java-v2/actions/runs/18512371544/job/52755900802?pr=2)

Considerations: 
This is not supposed to be the "end all be all" tool for catching these potential runtime violations between mixed versions. It's supposed to be a first line of defense for flagging changes with potentially risky elements. Essentially is a very blunt instrument that will require us to do some manual validation after it flags potential violations. 

We can always disable this if its more annoying than helpful.

RanVaknin committed:
Add mixed version compatibility detection workflow

- Detects changes to base classes (AwsRequest, AwsResponse, SdkPojo, etc.)
- Requires manual review via 'mixed-version-compatibility-reviewed' label
- Prevents merge until team approves compatibility impact

RanVaknin committed:
Improve mixed version detection: filter false positives

- Filter out comments containing 'public.*(' patterns
- Filter out string literals with 'public.*(' patterns
- Filter out javadoc examples with 'public.*(' patterns
- Reduces noise while maintaining detection accuracy

RanVaknin committed:
Fix false positive filtering for block comments

Add filtering for block comments that start with /*
This should catch patterns like: /* Example usage: public void method() */

alextwoods commented on .github/workflows/mixed-version-compatibility-review.yml:
```diff
@@ -0,0 +1,93 @@
+          echo "$CHANGED_BASE_FILES"
+          
+          # Look for new public methods in the changed base class files
+          # Filter out obvious false positives: comments, string literals, javadoc
```
I'm not sure I follow this comment - it looks like below we aren't filtering out comments,javadocs, ect - instead we're just checking for probably new public methods?

alextwoods commented on .github/workflows/mixed-version-compatibility-review.yml:
```diff
@@ -0,0 +1,93 @@
+            grep -v '^+[[:space:]]*//.*public.*(' | \
+            grep -v '^+.*".*public.*(' | \
+            grep -v '^+[[:space:]]*\*.*public.*(' | \
+            grep -v '^+[[:space:]]*/\*.*public.*(' || true)
```
Not sure if this makes sense or not - but could these regex be combined into a single more flexible regex? or are they really different cases?

RanVaknin replied:
mmm I feel like that would be harder to read and maintain. for example, i ended up adding more "rules" to this exclusion list. I think making this one big regex would be very difficult to iterate / separate. 

Did I misunderstand your comment? Can you give an example of what you mean?

RanVaknin replied:
Done 👍

RanVaknin committed:
Address feedback: show only changed class names in error message

- Extract class names from actually changed files, not all possible files
- Makes error message dynamic and specific to detected changes
- Addresses maintainability concern from code review

RanVaknin commented on .github/workflows/mixed-version-compatibility-review.yml:
```diff
@@ -0,0 +1,93 @@
+        run: |
+          echo "::error::Mixed version compatibility risk detected!"
+          echo "::error::Changes were made to base classes that generated service code implements:"
+          echo "::error::- AwsRequest, AwsResponse, AwsResponseMetadata, AwsServiceException, SdkPojo"
```
ack. fixed. now the workflow will specify which exact file had the risk detected.

Test: https://github.com/RanVaknin/aws-sdk-java-v2/actions/runs/18543837327/job/52857526599?pr=6

RanVaknin committed:
Add comments, expand exclusion of patterns
